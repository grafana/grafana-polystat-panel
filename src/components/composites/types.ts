import { SelectableValue } from '@grafana/data';

export interface CompositeMember {}

export const DisplayModes: SelectableValue[] = [
  { value: 'all', label: 'Show All' },
  { value: 'triggered', label: 'Show Triggered' },
];

export interface CompositeMetric {
  seriesMatch: string;
  compositeMatch?: CompositeItemType[];
  alias?: string;
  ID?: string;
  order: number;
}

export interface CompositeItemType {
  name: string;
  label: string;
  order: number;
  templatedName: string;
  isTemplated: boolean;
  displayMode: string;
  enabled: boolean;
  showName: boolean;
  showValue: boolean;
  showComposite: boolean;
  showMembers: boolean;
  metrics: CompositeMetric[];
  clickThrough: string | '';
  clickThroughSanitize: boolean;
  clickThroughOpenNewTab: boolean;
}

export interface CompositeItemTracker {
  composite: CompositeItemType;
  order: number;
  ID: string;
}

export interface CompositeItemProps {
  composite: CompositeItemType;
  ID: string;
  enabled: boolean;
  setter: any;
  remover: any;
  moveUp: any;
  moveDown: any;
  createDuplicate: any;
}

export interface CompositeMetricItemProps {
  metric: CompositeMetric;
  index: number;
  disabled: boolean;
  removeMetric: any;
  updateMetric: any;
  updateMetricAlias: any;
}
