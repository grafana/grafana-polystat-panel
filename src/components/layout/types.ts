
export interface LayoutPoint {
  x: number,
  y: number,
}
