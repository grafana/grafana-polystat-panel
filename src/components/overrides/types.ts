import { PolystatThreshold } from 'components/thresholds/types';

export interface OverrideItemProps {
  override: OverrideItemType;
  ID: string;
  enabled: boolean;
  setter: any;
  remover: any;
  moveUp: any;
  moveDown: any;
  createDuplicate: any;
}

export interface OverrideItemType {
  label: string;
  metricName: string;
  alias: string;
  thresholds: PolystatThreshold[];
  colors: string[];
  unitFormat: string;
  decimals: string;
  scaledDecimals: number | null;
  enabled: boolean;
  operatorName: string; // mean/min/max etc
  prefix: string;
  suffix: string;
  clickThrough: string | '';
  clickThroughSanitize: boolean;
  clickThroughOpenNewTab: boolean;
  order: number;
}

export interface OverrideItemTracker {
  override: OverrideItemType;
  order: number;
  ID: string;
}
