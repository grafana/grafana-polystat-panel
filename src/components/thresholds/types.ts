import { SelectableValue } from '@grafana/data';

export interface PolystatThreshold {
  color: string;
  state: number;
  value: number;
}

export interface ThresholdItemTracker {
  threshold: PolystatThreshold;
  order: number;
  ID: string;
}

export const ThresholdStates: SelectableValue[] = [
  { value: 0, label: 'ok' },
  { value: 1, label: 'warning' },
  { value: 2, label: 'critical' },
  { value: 3, label: 'custom' },
];
