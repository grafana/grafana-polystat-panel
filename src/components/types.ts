import { SelectableValue } from '@grafana/data';
import { CompositeItemType } from 'components/composites/types';
import { OverrideItemType } from './overrides/types';
import { PolystatThreshold } from './thresholds/types';

export interface PolystatOptions {
  autoSizeColumns: boolean;
  autoSizeRows: boolean;
  autoSizePolygons: boolean;
  ellipseCharacters: number;
  ellipseEnabled: boolean;
  globalAutoScaleFonts: boolean;
  globalClickthrough: string;
  globalClickthroughNewTabEnabled: boolean;
  globalClickthroughSanitizedEnabled: boolean;
  globalDecimals: number;
  globalDisplayMode: string;
  globalDisplayTextTriggeredEmpty: string;
  globalFillColor: string;
  globalFontSize: number;
  globalGradientsEnabled: boolean;
  globalOperator: string;
  globalPolygonBorderSize: number;
  globalPolygonBorderColor: string;
  globalPolygonSize: number;
  globalRegexPattern: string;
  globalShape: PolygonShapes;
  globalShowValueEnabled: boolean;
  globalTextFontColor: string;
  globalTextFontAutoColor: string;
  globalTextFontAutoColorEnabled: boolean;
  globalThresholdsConfig: PolystatThreshold[];
  globalTooltipsEnabled: boolean;
  globalTooltipsShowTimestampEnabled: boolean;
  globalUnitFormat: string;
  layoutDisplayLimit: number;
  layoutNumColumns: number;
  layoutNumRows: number;
  processedData?: PolystatModel[] | null;
  panelHeight: any;
  panelWidth: any;
  panelId: number;
  radius: number | null;
  renderTime?: Date;
  sortByField: string;
  sortByDirection: number;
  overrideConfig: {
    overrides: OverrideItemType[];
  };
  compositeConfig: {
    animationSpeed: string;
    composites: CompositeItemType[];
    enabled: boolean;
  };
  tooltipPrimarySortDirection: number;
  tooltipPrimarySortByField: string;
  tooltipSecondarySortDirection: number;
  tooltipSecondarySortByField: string;
  tooltipDisplayMode: string;
  tooltipDisplayTextTriggeredEmpty: string;
}

export interface PolystatModel {
  displayMode?: string;
  thresholdLevel?: number; // 0 = ok, 1 = warn, 2 = crit, 3 = unknown (same as sensu)
  value: number;
  valueFormatted: string;
  valueRounded: number;
  stats: any;
  name: string;
  displayName: string; // Used for composites
  timestamp: number;
  prefix: string;
  suffix: string;
  color: string;
  clickThrough: string;
  operatorName: string;
  newTabEnabled: boolean;
  sanitizedURL: string;
  sanitizeURLEnabled: boolean;
  showName: boolean;
  showValue: boolean;
  isComposite: boolean;
  members: PolystatModel[];
  triggerCache?: any; // holds animation frame info
}

export enum PolygonShapes {
  HEXAGON_POINTED_TOP = 'hexagon_pointed_top',
  CIRCLE = 'circle',
  SQUARE = 'square',
}

export const PolygonNamedShapes = [
  { value: 'hexagon_pointed_top', label: 'Hexagon Pointed Top' },
  { value: 'circle', label: 'Circle' },
  { value: 'square', label: 'Square' },
];

export interface PolystatDiameters {
  diameterX: number;
  diameterY: number;
}

export const OperatorOptions: SelectableValue[] = [
  { value: 'mean', label: 'Mean (avg)' },
  { value: 'sum', label: 'Sum' },
  { value: 'min', label: 'Min' },
  { value: 'max', label: 'Max' },
  { value: 'last', label: 'Last' },
  { value: 'lastNotNull', label: 'Last Not Null' },
  { value: 'first', label: 'First' },
  { value: 'firstNotNull', label: 'First Not Null' },
  { value: 'count', label: 'Count' },
  { value: 'allIsNull', label: 'All Is Null (boolean)' },
  { value: 'allIsZero', label: 'All Is Zero (boolean)' },
  { value: 'delta', label: 'Delta' },
  { value: 'diff', label: 'Difference' },
  { value: 'diffperc', label: 'Difference (Percent)' },
  { value: 'last_time', label: 'Time of Last Point' },
  { value: 'logmin', label: 'Log Min' },
  { value: 'name', label: 'Name' },
  { value: 'nonNullCount', label: 'Non Null Count' },
  { value: 'previousDeltaUp', label: 'Previous Delta Up' },
  { value: 'range', label: 'Range' },
  { value: 'step', label: 'Step' },
];

export const SortOptions = [
  { value: 0, label: 'Disabled' },
  { value: 1, label: 'Alphabetical (asc)' },
  { value: 2, label: 'Alphabetical (desc)' },
  { value: 3, label: 'Numerical (asc)' },
  { value: 4, label: 'Numerical (desc)' },
  { value: 5, label: 'Alphabetical (case-insensitive, asc)' },
  { value: 6, label: 'Alphabetical (case-insensitive, desc)' },
];

export const SortOptionOperators = ['disabled', 'asc', 'desc', 'asc', 'desc', 'iasc', 'idesc'];

export const SortByFieldOptions = [
  { value: 'name', label: 'Name' },
  { value: 'thresholdLevel', label: 'Threshold Level' },
  { value: 'value', label: 'Value' },
];

export const DisplayModes = [
  { value: 'all', label: 'Show All' },
  { value: 'triggered', label: 'Show Triggered' },
];

export const AnimationModes = [
  { value: 'all', text: 'Show All' },
  { value: 'triggered', text: 'Show Triggered' },
];
