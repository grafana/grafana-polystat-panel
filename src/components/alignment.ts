import { PolygonShapes } from "./types";

export const GetAlignments = (
  shape: PolygonShapes,
  diameterX: number,
  diameterY: number,
  textAreaHeight: number,
  activeValueFontSize: number,
  activeLabelFontSize: number,
  activeTimestampFontSize: number,
  showTimestampEnabled: boolean,
) => {
  let valueWithLabelTextAlignment = textAreaHeight / 2 / 2 + activeValueFontSize / 2;
  let valueOnlyTextAlignment = activeValueFontSize / 2;
  let labelWithValueTextAlignment = -(textAreaHeight / 2 / 2) + activeLabelFontSize / 2;
  let labelOnlyTextAlignment = activeLabelFontSize / 2;
  let labelTextAlignmentX = 0;
  let labelValueAlignmentX = 0;
  let valueWithTimestampAlignment = valueWithLabelTextAlignment / 2;
  let timestampAlignment = textAreaHeight * 0.33 / 2 + activeTimestampFontSize / 2;
  switch (shape) {
    case PolygonShapes.HEXAGON_POINTED_TOP:
      // offset when only showing label
      labelOnlyTextAlignment = activeLabelFontSize * 0.37;
      if (showTimestampEnabled && activeTimestampFontSize > 0) {
        // adjust value down
        valueWithLabelTextAlignment = textAreaHeight * 0.67 / 2 + activeValueFontSize / 2;
      }
      break;
    case PolygonShapes.CIRCLE:
      // offset when only showing label
      labelOnlyTextAlignment = activeLabelFontSize * 0.37;
      break;
    case PolygonShapes.SQUARE:
      // square is "centered" at top left, not the center

      // compute alignment for each text element, base coordinate is in the top left corner (text is anchored at its bottom):
      // - Value text (bottom text) will be aligned (positively i.e. lower) in the middle of the bottom half of the text area
      // - Label text (top text) will be aligned in the middle of the top half of the text area
      valueWithLabelTextAlignment = diameterY / 1.5 + activeValueFontSize / 2;
      valueOnlyTextAlignment = diameterY / 2 + activeLabelFontSize * 0.37;
      labelWithValueTextAlignment = diameterY / 4 + activeLabelFontSize / 2;
      // alignment is equal to the half of height plus a fraction of the fontSize
      labelOnlyTextAlignment = diameterY / 2 + activeLabelFontSize * 0.37;
      //
      labelTextAlignmentX = diameterX / 2;
      labelValueAlignmentX = diameterX / 2;
      break;
  }
  return {
    valueWithLabelTextAlignment,
    valueOnlyTextAlignment,
    valueWithTimestampAlignment,
    labelWithValueTextAlignment,
    labelOnlyTextAlignment,
    labelTextAlignmentX,
    labelValueAlignmentX,
    timestampAlignment,
  };
};
