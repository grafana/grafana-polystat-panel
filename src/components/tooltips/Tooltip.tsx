import React from 'react';
import { css } from '@emotion/css';
import { orderBy as lodashOrderBy } from 'lodash';
import { useStyles2 } from '@grafana/ui';
import { GrafanaTheme2, dateTimeFormatWithAbbrevation } from '@grafana/data';
import Tippy from '@tippyjs/react';
import { followCursor } from 'tippy.js';
import { PolystatModel, SortOptions } from '../types';

export interface TooltipProps {
  data: PolystatModel | undefined;
  valueEnabled: boolean;
  tooltipColumnHeadersEnabled: boolean;
  followMouse?: boolean;
  reference: any;
  visible: boolean;
  renderTime: Date;
  showTime: boolean;
  primarySortByField: string;
  primarySortDirection: number;
  secondarySortByField: string;
  secondarySortDirection: number;
  displayMode: string;
  tooltipDisplayTextTriggeredEmpty: string;
}

export const Tooltip = ({
  data,
  renderTime,
  showTime,
  valueEnabled,
  tooltipColumnHeadersEnabled,
  reference,
  visible,
  followMouse,
  primarySortByField,
  primarySortDirection,
  secondarySortByField,
  secondarySortDirection,
  displayMode,
  tooltipDisplayTextTriggeredEmpty,
}: TooltipProps) => {
  const styles = useStyles2(getTooltipStyles);

  /* the name of the composite is shown at the top */
  const getCompositeHeader = (data: PolystatModel | null) => {
    if (data && data.members && data.members.length === 0) {
      return '';
    }
    return (
      <tr>
        <th className={styles.tooltipCompositeHeading} colSpan={2}>
          {data!.displayName}
        </th>
      </tr>
    );
  };

  const multiSort = (
    data: PolystatModel,
    primarySortDirection: number,
    primarySortByField: string,
    secondarySortDirection: number,
    secondarySortByField: string
  ) => {
    let sortSettings = getSortDirection(SortOptions[primarySortDirection].value);
    const pUseLowercase = sortSettings.lowerCase;
    const pDirection = sortSettings.direction as any;
    sortSettings = getSortDirection(SortOptions[secondarySortDirection].value);
    const sDirection = sortSettings.direction as any;
    const sUseLowercase = sortSettings.lowerCase;

    let pSortFunction = primarySortByField as any;
    if (pUseLowercase) {
      pSortFunction = (item: any) => {
        let val = item[primarySortByField];
        if (typeof val !== 'number') {
          val = val.toLowerCase();
        }
        return val;
      };
    }
    let sSortFunction = secondarySortByField as any;
    if (sUseLowercase) {
      sSortFunction = (item: any) => {
        let val = item[secondarySortByField];
        if (typeof val !== 'number') {
          val = val.toLowerCase();
        }
        return val;
      };
    }
    const sortedMembers = lodashOrderBy(data.members, [pSortFunction, sSortFunction], [pDirection, sDirection]);
    return sortedMembers;
  };

  const getTriggeredCount = (data: PolystatModel) => {
    let triggered = 0;
    if (data.thresholdLevel !== 0) {
      triggered++;
    }
    for (let j = 0; j < data.members.length; j++) {
      if (data.members[j].thresholdLevel !== 0) {
        triggered++;
      }
    }
    return triggered;
  };
  const filterTriggered = (items: PolystatModel) => {
    const triggerCount = getTriggeredCount(items);
    if (triggerCount > 0) {
      for (let i = 0; i < items.members.length; i++) {
        if (items.members[i].thresholdLevel === 0) {
          items.members.splice(i, 1);
        }
      }
      return items;
    } else {
      return null;
    }
  };
  const getCompositeMetrics = (data: PolystatModel | null) => {
    let dataToSort = data;
    if (dataToSort && data && displayMode === 'triggered') {
      dataToSort = filterTriggered(data);
    }
    if (dataToSort === null) {
      return <>{tooltipDisplayTextTriggeredEmpty}</>;
    }
    const sortedMembers = multiSort(
      dataToSort,
      primarySortDirection,
      primarySortByField,
      secondarySortDirection,
      secondarySortByField
    );
    return sortedMembers.map((item: PolystatModel, index: number) => {
      return (
        <tr key={index} style={{ color: item.color }}>
          <td className={styles.tooltipName}>{item.displayName}</td>
          {valueEnabled && <td className={styles.tooltipValue}>{item.valueFormatted}</td>}
        </tr>
      );
    });
  };

  const generateContent = () => {
    return (
      <table className={styles.tooltipTable}>
        <thead>
          {data && data.isComposite && getCompositeHeader(data)}
          <tr>
            {tooltipColumnHeadersEnabled && <th className={styles.tooltipNameHeading}>Name</th>}
            {tooltipColumnHeadersEnabled && valueEnabled && <th className={styles.tooltipValueHeading}>Value</th>}
          </tr>
        </thead>
        <tfoot>
          {showTime && (
            <tr>
              <td className={styles.tooltipTime} colSpan={2}>
                {dateTimeFormatWithAbbrevation(renderTime)}
              </td>
            </tr>
          )}
        </tfoot>
        <tbody>
          {data && data.isComposite ? (
            getCompositeMetrics(data)
          ) : (
            <tr style={{ color: data!.color }}>
              <td className={styles.tooltipName}>{data!.displayName}</td>
              {valueEnabled && <td className={styles.tooltipValue}>{data!.valueFormatted}</td>}
            </tr>
          )}
        </tbody>
      </table>
    );
  };
  return (
    <Tippy
      ref={reference}
      visible={visible}
      content={generateContent()}
      followCursor={!!followMouse}
      plugins={[followCursor]}
      animation={false}
      className={styles.root}
      placement="auto"
      maxWidth={500}
    />
  );
};

const getTooltipStyles = (theme: GrafanaTheme2) => {
  return {
    root: css`
      border-radius: ${theme.shape.borderRadius(1)};
      background-color: ${theme.colors.background.primary};
      box-shadow: ${theme.shadows.z2};
      padding: ${theme.spacing(1)};
    `,
    tooltipTable: css`
      width: 100%;
      color: ${theme.colors.text.primary};
      height: auto;
      padding: 10px;
    `,
    tooltipTime: css`
      text-align: center;
      color: ${theme.colors.text.primary};
    `,
    tooltipNameHeading: css`
      text-align: left;
      color: ${theme.colors.text.primary};
    `,
    tooltipValueHeading: css`
      text-align: right;
      color: ${theme.colors.text.primary};
    `,
    tooltipCompositeHeading: css`
      text-align: center;
      color: ${theme.colors.text.primary};
    `,
    tooltipName: css`
      text-align: left;
    `,
    tooltipValue: css`
      text-align: right;
      padding-left: 15px;
    `,
  };
};

const getSortDirection = (val: number) => {
  switch (val) {
    case 0:
      return { direction: null, lowerCase: false };
    case 1:
      return { direction: 'asc', lowerCase: false };
    case 2:
      return { direction: 'desc', lowerCase: false };
    case 3:
      return { direction: 'asc', lowerCase: false };
    case 4:
      return { direction: 'desc', lowerCase: false };
    case 5:
      return { direction: 'asc', lowerCase: true };
    case 6:
      return { direction: 'desc', lowerCase: true };
    default:
      return { direction: 'asc', lowerCase: false };
  }
};
