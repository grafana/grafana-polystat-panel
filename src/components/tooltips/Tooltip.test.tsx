import React, { createRef } from 'react';
import { render, screen } from '@testing-library/react';

import { SortOptions } from 'components/types';
import { ApplyComposites } from 'data/composite_processor';

import { Tooltip, TooltipProps } from './Tooltip';

import {
  modelA,
  modelB,
  modelC,
  numericalModelA,
  numericalModelB,
  numericalModelC,
  casedModelA,
  casedModelB,
  casedModelC,
} from '../../__mocks__/models/models';
import { compositeA, compositeB, compositeC } from '../../__mocks__/models/composites';

describe('Test Tooltips', () => {
  const aRef = createRef();
  var renderTime: Date;
  var props: TooltipProps;
  beforeAll(() => {
    renderTime = new Date('01 October 2022 10:28 UTC');
    props = {
      data: undefined,
      valueEnabled: true,
      reference: aRef,
      followMouse: false,
      visible: true,
      renderTime: renderTime,
      showTime: true,
      primarySortByField: 'threshold',
      primarySortDirection: 0,
      secondarySortByField: 'value',
      secondarySortDirection: 0,
      displayMode: 'all',
      tooltipDisplayTextTriggeredEmpty: 'EMPTY',
    };
  });

  /*
  beforeEach(() => {
  });
  */

  describe('Tooltip Sorting', () => {
    it('returns unsorted metrics', () => {
      const applied = ApplyComposites([compositeA], [modelA, modelB, modelC], (val) => val);
      const props: TooltipProps = {
        data: applied[0],
        valueEnabled: true,
        reference: aRef,
        followMouse: false,
        visible: true,
        renderTime: renderTime,
        showTime: true,
        primarySortByField: 'name',
        primarySortDirection: SortOptions[0].value,
        secondarySortByField: 'value',
        secondarySortDirection: SortOptions[0].value,
        displayMode: 'all',
        tooltipDisplayTextTriggeredEmpty: 'EMPTY',
      };
      render(<Tooltip reference={aRef} {...props} />);
      const rows = screen.getAllByRole('row');
      expect(rows.length).toBe(6);
      expect(rows[0].innerHTML).toContain('composite-a');
      expect(rows[2].innerHTML).toContain('2022-10-01 10:28:00');
      expect(rows[3].innerHTML).toContain('A-series');
      expect(rows[4].innerHTML).toContain('B-series');
      expect(rows[5].innerHTML).toContain('C-series');
    });
    it('returns primary sorted metrics: case sensitive ascending', () => {
      const applied = ApplyComposites([compositeA], [modelA, modelB, modelC], (val) => val);
      const props: TooltipProps = {
        data: applied[0],
        valueEnabled: true,
        reference: aRef,
        followMouse: false,
        visible: true,
        renderTime: renderTime,
        showTime: true,
        primarySortByField: 'name',
        primarySortDirection: SortOptions[1].value,
        secondarySortByField: 'value',
        secondarySortDirection: SortOptions[0].value,
        displayMode: 'all',
        tooltipDisplayTextTriggeredEmpty: 'EMPTY',
      };
      render(<Tooltip reference={aRef} {...props} />);
      const rows = screen.getAllByRole('row');
      expect(rows.length).toBe(6);
      expect(rows[0].innerHTML).toContain('composite-a');
      expect(rows[2].innerHTML).toContain('2022-10-01 10:28:00');
      expect(rows[3].innerHTML).toContain('A-series');
      expect(rows[4].innerHTML).toContain('B-series');
      expect(rows[5].innerHTML).toContain('C-series');
    });

    it('returns primary sorted metrics: case sensitive descending', () => {
      const applied = ApplyComposites([compositeA], [modelA, modelB, modelC], (val) => val);
      const props: TooltipProps = {
        data: applied[0],
        valueEnabled: true,
        reference: aRef,
        followMouse: false,
        visible: true,
        renderTime: renderTime,
        showTime: true,
        primarySortByField: 'name',
        primarySortDirection: SortOptions[2].value,
        secondarySortByField: 'value',
        secondarySortDirection: SortOptions[0].value,
        displayMode: 'all',
        tooltipDisplayTextTriggeredEmpty: 'EMPTY',
      };
      render(<Tooltip reference={aRef} {...props} />);
      const rows = screen.getAllByRole('row');
      expect(rows.length).toBe(6);
      expect(rows[0].innerHTML).toContain('composite-a');
      expect(rows[2].innerHTML).toContain('2022-10-01 10:28:00');
      expect(rows[3].innerHTML).toContain('C-series');
      expect(rows[4].innerHTML).toContain('B-series');
      expect(rows[5].innerHTML).toContain('A-series');
    });

    it('returns primary sorted metrics: numerical ascending', () => {
      let internalData = [numericalModelA, numericalModelB, numericalModelC];
      internalData = ApplyComposites([compositeC], internalData, (val) => val);
      const props: TooltipProps = {
        data: internalData[0],
        valueEnabled: true,
        reference: aRef,
        followMouse: false,
        visible: true,
        renderTime: renderTime,
        showTime: true,
        primarySortByField: 'name',
        primarySortDirection: SortOptions[3].value,
        secondarySortByField: 'value',
        secondarySortDirection: SortOptions[0].value,
        displayMode: 'all',
        tooltipDisplayTextTriggeredEmpty: 'EMPTY',
      };
      render(<Tooltip reference={aRef} {...props} />);
      const rows = screen.getAllByRole('row');
      expect(rows.length).toBe(6);
      expect(rows[0].innerHTML).toContain('composite-numerical');
      expect(rows[2].innerHTML).toContain('2022-10-01 10:28:00');
      expect(rows[3].innerHTML).toContain('01');
      expect(rows[4].innerHTML).toContain('02');
      expect(rows[5].innerHTML).toContain('03');
    });

    it('returns primary sorted metrics: numerical descending', () => {
      let internalData = [numericalModelA, numericalModelB, numericalModelC];
      internalData = ApplyComposites([compositeC], internalData, (val) => val);
      const props: TooltipProps = {
        data: internalData[0],
        valueEnabled: true,
        reference: aRef,
        followMouse: false,
        visible: true,
        renderTime: renderTime,
        showTime: true,
        primarySortByField: 'name',
        primarySortDirection: SortOptions[4].value,
        secondarySortByField: 'value',
        secondarySortDirection: SortOptions[0].value,
        displayMode: 'all',
        tooltipDisplayTextTriggeredEmpty: 'EMPTY',
      };
      render(<Tooltip reference={aRef} {...props} />);
      const rows = screen.getAllByRole('row');
      expect(rows.length).toBe(6);
      expect(rows[0].innerHTML).toContain('composite-numerical');
      expect(rows[2].innerHTML).toContain('2022-10-01 10:28:00');
      expect(rows[3].innerHTML).toContain('03');
      expect(rows[4].innerHTML).toContain('02');
      expect(rows[5].innerHTML).toContain('01');
    });

    it('returns primary sorted metrics: case insensitive ascending', () => {
      const applied = ApplyComposites([compositeA], [casedModelA, casedModelB, casedModelC], (val) => val);
      const props: TooltipProps = {
        data: applied[0],
        valueEnabled: true,
        reference: aRef,
        followMouse: false,
        visible: true,
        renderTime: renderTime,
        showTime: true,
        primarySortByField: 'name',
        primarySortDirection: SortOptions[5].value,
        secondarySortByField: 'value',
        secondarySortDirection: SortOptions[0].value,
        displayMode: 'all',
        tooltipDisplayTextTriggeredEmpty: 'EMPTY',
      };
      render(<Tooltip reference={aRef} {...props} />);
      const rows = screen.getAllByRole('row');
      expect(rows.length).toBe(6);
      expect(rows[0].innerHTML).toContain('composite-a');
      expect(rows[2].innerHTML).toContain('2022-10-01 10:28:00');
      expect(rows[3].innerHTML).toContain('series-a1');
      expect(rows[4].innerHTML).toContain('series-A2');
      expect(rows[5].innerHTML).toContain('series-a3');
    });

    it('returns primary sorted metrics: case insensitive descending', () => {
      const applied = ApplyComposites([compositeA], [casedModelA, casedModelB, casedModelC], (val) => val);
      const props: TooltipProps = {
        data: applied[0],
        valueEnabled: true,
        reference: aRef,
        followMouse: false,
        visible: true,
        renderTime: renderTime,
        showTime: true,
        primarySortByField: 'name',
        primarySortDirection: SortOptions[6].value,
        secondarySortByField: 'value',
        secondarySortDirection: SortOptions[0].value,
        displayMode: 'all',
        tooltipDisplayTextTriggeredEmpty: 'EMPTY',
      };
      render(<Tooltip reference={aRef} {...props} />);
      const rows = screen.getAllByRole('row');
      expect(rows.length).toBe(6);
      expect(rows[0].innerHTML).toContain('composite-a');
      expect(rows[2].innerHTML).toContain('2022-10-01 10:28:00');
      expect(rows[3].innerHTML).toContain('series-a3');
      expect(rows[4].innerHTML).toContain('series-A2');
      expect(rows[5].innerHTML).toContain('series-a1');
    });
  });

  describe('Tooltip Generation', () => {
    it('returns tooltip for single metric', () => {
      const applied = ApplyComposites([compositeA], [modelA], (val) => val);
      const props: TooltipProps = {
        data: applied[0],
        valueEnabled: true,
        reference: aRef,
        followMouse: false,
        visible: true,
        renderTime: renderTime,
        showTime: true,
        primarySortByField: 'name',
        primarySortDirection: SortOptions[0].value,
        secondarySortByField: 'value',
        secondarySortDirection: SortOptions[0].value,
        displayMode: 'all',
        tooltipDisplayTextTriggeredEmpty: 'EMPTY',
      };

      render(<Tooltip reference={aRef} {...props} />);
      expect(screen.getByRole('table')).toMatchSnapshot();
    });
    it('returns tooltip for composite metric with one metric', () => {
      const applied = ApplyComposites([compositeA], [modelA], (val) => val);
      props.data = applied[0];
      render(<Tooltip reference={aRef} {...props} />);
      expect(screen.getByRole('table')).toMatchSnapshot();
    });
    it('returns tooltip for composite metric with two metrics', () => {
      const applied = ApplyComposites([compositeA], [modelA, modelB], (val) => val);
      props.data = applied[0];
      render(<Tooltip reference={aRef} {...props} />);
      expect(screen.getByRole('table')).toMatchSnapshot();
    });
    it('returns tooltip for composite metric with three metrics', () => {
      const applied = ApplyComposites([compositeA], [modelA, modelB, modelC], (val) => val);
      props.data = applied[0];
      render(<Tooltip reference={aRef} {...props} />);
      expect(screen.getByRole('table')).toMatchSnapshot();
    });
    it('returns tooltip for composite metric with one triggered metric', () => {
      modelA.thresholdLevel = 0;
      modelB.thresholdLevel = 1;
      modelC.thresholdLevel = 0;
      const applied = ApplyComposites([compositeB], [modelA, modelB, modelC], (val) => val);
      props.data = applied[0];
      // the displayMode comes from the applied composite, in this case there is one triggered metric
      props.displayMode = applied[0].displayMode;
      render(<Tooltip reference={aRef} {...props} />);
      const rows = screen.getAllByRole('row');
      expect(rows.length).toBe(4);
      expect(rows[0].innerHTML).toContain('composite-b');
      expect(rows[2].innerHTML).toContain('2022-10-01 10:28:00');
      expect(rows[3].innerHTML).toContain('B-series');
    });
  });
});
