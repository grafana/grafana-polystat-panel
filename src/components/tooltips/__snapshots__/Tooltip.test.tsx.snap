// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Tooltips Tooltip Generation returns tooltip for composite metric with one metric 1`] = `
<table
  class="css-9iiucf"
>
  <thead>
    <tr>
      <th
        class="css-cxgyzm"
        colspan="2"
      >
        composite-a
      </th>
    </tr>
    <tr>
      <th
        class="css-rmdaym"
      >
        Name
      </th>
      <th
        class="css-dp6ifn"
      >
        Value
      </th>
    </tr>
  </thead>
  <tfoot>
    <tr>
      <td
        class="css-cxgyzm"
        colspan="2"
      >
        2022-10-01 10:28:00 
      </td>
    </tr>
  </tfoot>
  <tbody>
    <tr
      style="color: rgb(10, 85, 161);"
    >
      <td
        class="css-1bpvq0r"
      >
        A-series02
      </td>
      <td
        class="css-tydf9p"
      >
        210.00 MB/s
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Test Tooltips Tooltip Generation returns tooltip for composite metric with three metrics 1`] = `
<table
  class="css-9iiucf"
>
  <thead>
    <tr>
      <th
        class="css-cxgyzm"
        colspan="2"
      >
        composite-a
      </th>
    </tr>
    <tr>
      <th
        class="css-rmdaym"
      >
        Name
      </th>
      <th
        class="css-dp6ifn"
      >
        Value
      </th>
    </tr>
  </thead>
  <tfoot>
    <tr>
      <td
        class="css-cxgyzm"
        colspan="2"
      >
        2022-10-01 10:28:00 
      </td>
    </tr>
  </tfoot>
  <tbody>
    <tr
      style="color: rgb(10, 85, 161);"
    >
      <td
        class="css-1bpvq0r"
      >
        A-series02
      </td>
      <td
        class="css-tydf9p"
      >
        210.00 MB/s
      </td>
    </tr>
    <tr
      style="color: rgb(10, 85, 161);"
    >
      <td
        class="css-1bpvq0r"
      >
        C-series01
      </td>
      <td
        class="css-tydf9p"
      >
        444.0000
      </td>
    </tr>
    <tr
      style="color: rgb(10, 85, 161);"
    >
      <td
        class="css-1bpvq0r"
      >
        B-series03
      </td>
      <td
        class="css-tydf9p"
      >
        510.0000
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Test Tooltips Tooltip Generation returns tooltip for composite metric with two metrics 1`] = `
<table
  class="css-9iiucf"
>
  <thead>
    <tr>
      <th
        class="css-cxgyzm"
        colspan="2"
      >
        composite-a
      </th>
    </tr>
    <tr>
      <th
        class="css-rmdaym"
      >
        Name
      </th>
      <th
        class="css-dp6ifn"
      >
        Value
      </th>
    </tr>
  </thead>
  <tfoot>
    <tr>
      <td
        class="css-cxgyzm"
        colspan="2"
      >
        2022-10-01 10:28:00 
      </td>
    </tr>
  </tfoot>
  <tbody>
    <tr
      style="color: rgb(10, 85, 161);"
    >
      <td
        class="css-1bpvq0r"
      >
        A-series02
      </td>
      <td
        class="css-tydf9p"
      >
        210.00 MB/s
      </td>
    </tr>
    <tr
      style="color: rgb(10, 85, 161);"
    >
      <td
        class="css-1bpvq0r"
      >
        B-series03
      </td>
      <td
        class="css-tydf9p"
      >
        510.0000
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Test Tooltips Tooltip Generation returns tooltip for single metric 1`] = `
<table
  class="css-9iiucf"
>
  <thead>
    <tr>
      <th
        class="css-cxgyzm"
        colspan="2"
      >
        composite-a
      </th>
    </tr>
    <tr>
      <th
        class="css-rmdaym"
      >
        Name
      </th>
      <th
        class="css-dp6ifn"
      >
        Value
      </th>
    </tr>
  </thead>
  <tfoot>
    <tr>
      <td
        class="css-cxgyzm"
        colspan="2"
      >
        2022-10-01 10:28:00 
      </td>
    </tr>
  </tfoot>
  <tbody>
    <tr
      style="color: rgb(10, 85, 161);"
    >
      <td
        class="css-1bpvq0r"
      >
        A-series02
      </td>
      <td
        class="css-tydf9p"
      >
        210.00 MB/s
      </td>
    </tr>
  </tbody>
</table>
`;
